### DISCUSSION TITLE
Shorten the path of the RPM-controlled per-build directory

### DISCUSSION THREAD

**Discussion by lantw44** [rpm-software-management/rpm#3438] [Q&A]
Is it possible to shorten the `<package>-<version>-build` part from a spec file?

I package Guix in Fedora Copr. To run Guix tests in the rpmbuild %check step, the full path of the working directory cannot be longer than 29 bytes. The default `/builddir/build/BUILD/guix-1.4.0` used by mock is already too long, so I have to change the default `test-tmp` directory used by Guix tests to `t` to save a few bytes in order to run tests.

Now the new default `/builddir/build/BUILD/guix-1.4.0-build/guix-1.4.0` is even longer than `/builddir/build/BUILD/guix-1.4.0`, so the `test-tmp` → `t` trick can't work. Is it possible to tell rpmbuild to use a shorter path than `guix-1.4.0-build`, or I should just `mktemp -d` and tell Guix to run tests in the temporary directory?

https://github.com/rpm-software-management/rpm/issues/2078
https://github.com/rpm-software-management/rpm/pull/2885

**Comment by pmatilai** [comment]
No there isn't. The precise location of that directory is rpm's business only.

What's this absurd 29 byte limit? I'd suggest patching *that* instead.

**Reply by lantw44** [reply]
Each Guix package directory includes a hash, so it is typical for a directory to take 50 bytes. For example, you can find `/gnu/store/4y5m9lb8k3qkb1y9m02sw9w9a6hacd16-bash-minimal-5.1.8` in its official binary tarball. The problem is that Guix uses this kind of long path in shebangs and Unix socket paths, so it can hit the limit if the test directory is too long.

It looks like moving the test directory to /tmp (using `mktemp -d`) with an autoconf cache variable is the only solution.

**Comment by voxik** [comment]
The path certainly become too long:

https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/FZSQW6CMUWBUGW6YVJ4IXPSVODKPXNR6/

**Reply by lantw44** [reply]
Yes, Guix has length checks. These lines are extracted from the build log on Fedora 40:
```
checking for the store directory... /gnu/store
checking the length of the installed socket file name... 30
checking for unit test root directory... (cached) /builddir/build/BUILD/guix-1.4.0/t
checking the length of the socket file name used in tests... 66
checking the length of a typical hash bang line... 73
checking the length of a hash bang line used in tests... 103
```
